File: clients/typescript/test/contract.test.ts
/**
 * Contract tests against a real, isolated StreamsForge engine (see engine-fixture.ts), run once
 * per transport -- the whole justification for the Transport interface (design doc §3.6/§8):
 * implementations that agree on every assertion are interchangeable, and one that drifts fails on
 * the same line the others pass.
 *
 * Owns: connect() + auth, subscribe -> snapshot -> replay, LATEST BY supersession over the wire,
 * GROUP BY aggregation over a derived LATEST BY table, and (once, transport-agnostic) validate +
 * ad-hoc SQL create/drop.
 */

import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { connect, type Client, type TransportName } from "../src/index.js";
import { bootEngine, preflightSkipReason, type Engine } from "./engine-fixture.js";

const skipReason = preflightSkipReason();
const describeOrSkip = skipReason ? describe.skip : describe;
if (skipReason) console.warn(`streamsforge contract tests: SKIPPED -- ${skipReason}`);

describeOrSkip("contract: isolated engine", () => {
  let engine: Engine;

  beforeAll(async () => {
    engine = await bootEngine();
  }, 180_000);

  afterAll(async () => {
    await engine?.stop();
  }, 20_000);

  const transports: TransportName[] = ["grpc", "signalr:ws", "signalr:sse", "signalr:lp"];

  for (const transportName of transports) {
    describe(transportName, () => {
      let client: Client;

      beforeAll(async () => {
        client = await connect({
          url: engine.baseUrl,
          grpc: engine.grpcTarget,
          user: engine.user,
          password: engine.password,
          transport: transportName,
        });
      }, 30_000);

      afterAll(async () => {
        await client?.close();
      });

      test("connects via the requested transport", () => {
        expect(client.transportName).toBe(transportName);
      });

      test("no delta lost across subscribe establishment (regression: subscribe must register before snapshot)", async () => {
        // Regression coverage for a bug found in this project's Kotlin client (cold Flow -- the
        // subscription didn't actually connect until collection began) and independently in a
        // narrower form in Python (socket eager, hub handshake lazy): Transport.subscribe() must
        // fully register with the server -- not just return an unstarted async iterable -- before
        // LiveTable's snapshot read fires, or a delta landing in that window is gone for good (a
        // LATEST BY row asserted once and never touched again gets no backfill to recover it).
        //
        // Firing the push and the table() open in the same tick exercises exactly that window:
        // under the old (buggy) `async function* subscribe()` shape, the generator's body --
        // including the real handshake -- didn't run until live-table.ts's first `.next()`, so
        // the snapshot read could fire (and for SignalR, reliably WOULD fire, since establishing
        // a fresh connection is far slower than one REST round trip) while the server still had
        // no idea we existed. This LATEST BY row is asserted exactly once, so there is nothing
        // downstream to self-heal a lost delta -- it either shows up, or it is gone.
        const tag = `race_${transportName.replace(/[^a-z0-9]/g, "_")}`;
        const [, table] = await Promise.all([
          client.push(engine.source, [{ trade_id: `${tag}-T1`, desk: "Rates", notional: 111 }]),
          client.table(engine.latestTable, { key: ["trade_id"], timeoutMs: 30_000 }),
        ]);
        try {
          await table.waitFor((rows) => rows.some((r) => r.trade_id === `${tag}-T1`), 10_000);
          expect(table.value("notional", { trade_id: `${tag}-T1` })).toBe(111);
        } finally {
          table.close();
        }
      }, 30_000);

      test("subscribe -> snapshot -> replay, LATEST BY supersession, GROUP BY aggregation", async () => {
        const tag = transportName.replace(/[^a-z0-9]/g, "_");

        // Subscribe to BOTH tables before pushing anything -- desk_totals is a table reading
        // ANOTHER table (GROUP BY over the LATEST BY view), and this engine's Rows/snapshot read
        // for that shape lags its own live dataflow by up to its persistence flush interval
        // (~2s, CLAUDE.md's "Flush interval in ms for Batched/FireAndForget" default): a snapshot
        // taken between the push and that flush can read stale/empty even though the live delta
        // stream has already emitted the update. Subscribing first sidesteps the race entirely
        // and is also the realistic usage pattern (subscribe once, watch data flow in) --
        // confirmed empirically that a live subscriber sees the update in ~150ms.
        await using latest = await client.table(engine.latestTable, { key: ["trade_id"], timeoutMs: 30_000 });
        await using agg = await client.table(engine.aggTable, { key: ["desk"], timeoutMs: 30_000 });

        await client.push(engine.source, [
          { trade_id: `${tag}-T1`, desk: "Rates", notional: 100 },
          { trade_id: `${tag}-T2`, desk: "Credit", notional: 40 },
        ]);
        await latest.waitFor((rows) => rows.filter((r) => String(r.trade_id).startsWith(tag)).length >= 2, 15_000);

        const t1 = latest.value("notional", { trade_id: `${tag}-T1` });
        expect(t1).toBe(100);

        // A LATEST BY update over the wire: the old row for T1 must be superseded, not duplicated.
        await client.push(engine.source, [{ trade_id: `${tag}-T1`, desk: "Rates", notional: 250 }]);
        await latest.waitFor((rows) => rows.some((r) => r.trade_id === `${tag}-T1` && r.notional === 250), 15_000);

        const mine = latest.rows.filter((r) => String(r.trade_id).startsWith(tag));
        expect(mine.length).toBe(2); // still exactly one row per trade_id -- no duplicate from the supersession

        // GROUP BY over the derived LATEST BY table must reflect the superseded value, not both.
        await agg.waitFor((rows) => {
          const rates = rows.find((r) => r.desk === "Rates");
          return rates !== undefined && Number(rates.total) >= 250;
        }, 15_000);
      }, 30_000);
    });
  }

  test("keyFields resolved from the engine (transport-agnostic, wishlist #18)", async () => {
    // client.table() with no `key` must read the table's own `keyFields` (GET /api/tables)
    // instead of the deleted hand-maintained map -- correct supersession (never a duplicate row)
    // is the observable proof the resolved key was actually used, not just that *a* key was, for
    // all three wire states: non-empty (LATEST BY), [] (global aggregate) and, implicitly, the
    // GROUP BY case shared with the existing suite above.
    const client = await connect({
      url: engine.baseUrl,
      grpc: engine.grpcTarget,
      user: engine.user,
      password: engine.password,
      transport: "signalr:ws",
    });
    try {
      const tag = "keyfields_engine";

      // LATEST BY -- non-empty keyFields (["trade_id"]).
      {
        await using latest = await client.table(engine.latestTable, { timeoutMs: 30_000 }); // no key=
        await client.push(engine.source, [{ trade_id: `${tag}-T1`, desk: "Rates", notional: 100 }]);
        await latest.waitFor((rows) => rows.some((r) => r.trade_id === `${tag}-T1`), 15_000);
        await client.push(engine.source, [{ trade_id: `${tag}-T1`, desk: "Rates", notional: 250 }]);
        await latest.waitFor((rows) => rows.some((r) => r.trade_id === `${tag}-T1` && r.notional === 250), 15_000);
        const mine = latest.rows.filter((r) => r.trade_id === `${tag}-T1`);
        expect(mine.length).toBe(1); // superseded, not duplicated
      }

      // GROUP BY -- non-empty keyFields (["desk"]).
      {
        const desk = `Desk-${tag}`;
        await using agg = await client.table(engine.aggTable, { timeoutMs: 30_000 }); // no key=
        await client.push(engine.source, [{ trade_id: `${tag}-A1`, desk, notional: 40 }]);
        await client.push(engine.source, [{ trade_id: `${tag}-A2`, desk, notional: 60 }]);
        await agg.waitFor((rows) => {
          const row = rows.find((r) => r.desk === desk);
          return row !== undefined && Number(row.total) >= 100;
        }, 15_000);
        expect(agg.rows.filter((r) => r.desk === desk).length).toBe(1);
      }

      // No GROUP BY at all -- keyFields is [] (a global aggregate: one row, one group). If the
      // resolver ever collapsed [] to null (whole-row identity) this table would grow a second
      // row on the next push instead of staying at exactly one.
      {
        await using globalAgg = await client.table(engine.globalAggTable, { timeoutMs: 30_000 }); // no key=
        await client.push(engine.source, [{ trade_id: `${tag}-G1`, desk: "Global", notional: 10 }]);
        await globalAgg.waitFor((rows) => rows.length >= 1, 15_000);
        await client.push(engine.source, [{ trade_id: `${tag}-G2`, desk: "Global", notional: 20 }]);
        await globalAgg.waitFor((rows) => rows.length === 1 && Number(rows[0]?.trade_count ?? 0) >= 2, 15_000);
        expect(globalAgg.rows.length).toBe(1);
      }
    } finally {
      await client.close();
    }
  }, 30_000);

  test("validate + ad-hoc SQL create/replace/drop (transport-agnostic REST)", async () => {
    const client = await connect({
      url: engine.baseUrl,
      grpc: engine.grpcTarget,
      user: engine.user,
      password: engine.password,
      transport: "signalr:ws",
    });
    try {
      const bad = await client.validate("SELECT this is not valid sql (((");
      expect(bad.ok).toBe(false);
      expect(bad.diagnostics.length).toBeGreaterThan(0);

      await using adhocTable = await client.sql(`SELECT desk, SUM(notional) AS total FROM ${engine.latestTable} GROUP BY desk`, {
        name: "contract test scratch",
        key: ["desk"],
        timeoutMs: 30_000,
      });
      const listed = await client.adhoc();
      expect(listed.some((t) => t.name === "adhoc_contract_test_scratch")).toBe(true);

      const dropped = await client.dropAdhoc("adhoc_contract_test_scratch");
      expect(dropped).toBe(true);
      void adhocTable; // already dropped server-side; local handle just needs to close cleanly
    } finally {
      await client.close();
    }
  }, 30_000);
});

File: clients/kotlin/src/test/kotlin/streamsforge/ContractTest.kt
package streamsforge

import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Timeout
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestMethodOrder
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import java.util.UUID
import kotlin.time.Duration.Companion.seconds

/**
 * Contract tests against a REAL, isolated StreamsForge instance ([EngineFixture]), parametrized
 * over every live transport (gRPC + SignalR) -- one set of assertions proving they're actually
 * interchangeable, per [TableTransport]'s whole reason for existing (design doc §3.6/§8). Ported
 * from the Python client's `tests/test_contract.py`.
 *
 * All tests share one engine instance ([EngineFixture] boots once per class) and its two fixture
 * tables -- [Order] pins `handshakeAndSnapshot` first, mirroring the Python suite's file-order
 * reliance on running before anything else has pushed a row into the shared `latest_table`
 * (JUnit5 does not otherwise guarantee method execution order).
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class ContractTest {
    private lateinit var handle: EngineFixture.Handle

    @BeforeAll
    fun startEngine() {
        val skipReason = EngineFixture.preconditionsOrSkipReason()
        assumeTrue(skipReason == null, skipReason)
        handle = EngineFixture.start()
    }

    @AfterAll
    fun stopEngine() {
        if (::handle.isInitialized) EngineFixture.stop(handle)
    }

    private suspend fun connectVia(transport: Transport): StreamsForgeClient {
        val client = StreamsForge.connect(
            url = EngineFixture.baseUrl,
            grpcTarget = EngineFixture.grpcTarget,
            user = EngineFixture.ADMIN_USER,
            password = EngineFixture.ADMIN_PASS,
            transport = transport,
        )
        assertEquals(transport.name.lowercase(), client.transportName)
        return client
    }

    @Order(1)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun handshakeAndSnapshot(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            // A freshly-imported table has no rows yet -- snapshot must succeed and be empty,
            // not error.
            val rows = sf.snapshot(EngineFixture.LATEST_TABLE)
            assertTrue(rows.isEmpty())
        }
    }

    @Order(2)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun pushThenLiveTableSeesIt(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            val tradeId = "t-${UUID.randomUUID().toString().take(8)}"
            val ack = sf.push(
                EngineFixture.SOURCE_NAME,
                listOf(mapOf("trade_id" to tradeId, "desk" to "Rates", "notional" to 100.0)),
            )
            assertEquals(1, ack.accepted)

            // REST ingest (used here under SIGNALR -- gRPC's bidi Ingest is synchronous
            // backpressure, so a GRPC-transport push is already committed by the time `ack`
            // returns) acks 202 once admitted, not once the row store is updated. Subscribing to
            // a FRESH table gets no backfill (design doc), so if this single push's own commit
            // hasn't landed yet, nothing will ever re-assert it and waitFor below would hang
            // until timeout. Wait for the row store itself to catch up first -- same technique
            // `tests/conftest.py`'s `test_reader_thread_reconnects_after_close` uses for the
            // identical race.
            waitForSnapshotToContain(sf, EngineFixture.LATEST_TABLE, tradeId)

            sf.table(EngineFixture.LATEST_TABLE, keyFields = listOf("trade_id")).use { t ->
                val rows = t.waitFor(45.seconds) { r -> r.any { it["trade_id"] == tradeId } }
                val row = rows.first { it["trade_id"] == tradeId }
                assertEquals("Rates", row["desk"])
                assertEquals(100.0, row["notional"])
            }
        }
    }

    private suspend fun waitForSnapshotToContain(sf: StreamsForgeClient, table: String, tradeId: String) {
        val deadline = kotlin.time.TimeSource.Monotonic.markNow() + 15.seconds
        while (sf.snapshot(table).none { it["trade_id"] == tradeId }) {
            if (deadline.hasPassedNow()) org.junit.jupiter.api.Assertions.fail<Unit>("engine snapshot never caught up with the earlier push")
            kotlinx.coroutines.delay(250)
        }
    }

    @Order(3)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun supersessionLatestBy(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            val tradeId = "t-${UUID.randomUUID().toString().take(8)}"
            sf.table(EngineFixture.LATEST_TABLE, keyFields = listOf("trade_id")).use { t ->
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to tradeId, "desk" to "Rates", "notional" to 100.0)))
                t.waitFor(45.seconds) { r -> r.any { it["trade_id"] == tradeId } }

                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to tradeId, "desk" to "Rates", "notional" to 250.0)))
                val rows = t.waitFor(45.seconds) { r ->
                    val match = r.filter { it["trade_id"] == tradeId }
                    match.size == 1 && match.first()["notional"] == 250.0
                }
                // LATEST BY: exactly one row for this trade_id, never two.
                assertEquals(1, rows.count { it["trade_id"] == tradeId })
            }
        }
    }

    @Order(4)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun globalAggregateReflectsPushes(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            val desk = "Desk-${UUID.randomUUID().toString().take(6)}"
            sf.table(EngineFixture.AGG_TABLE, keyFields = listOf("desk")).use { agg ->
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to "a-${UUID.randomUUID()}", "desk" to desk, "notional" to 40.0)))
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to "b-${UUID.randomUUID()}", "desk" to desk, "notional" to 60.0)))
                agg.waitFor(45.seconds) { r ->
                    val match = r.filter { it["desk"] == desk }
                    match.size == 1 && match.first()["total"] == 100.0
                }
            }
        }
    }

    @Order(5)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(30)
    fun ingestRowErrorsOnBadRow(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            // A string where a Double is declared fails coercion under this source's default
            // OnCoercionFailure -- a real rejection, not a lenient null-fill.
            val ex = runCatching {
                sf.push(
                    EngineFixture.SOURCE_NAME,
                    listOf(mapOf("trade_id" to "t-${UUID.randomUUID()}", "desk" to "Ops", "notional" to "not-a-number")),
                )
            }.exceptionOrNull()
            assertTrue(ex is IngestRejectedException, "expected IngestRejectedException, got $ex")
        }
    }

    @Order(6)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(30)
    fun validateRejectsBadSql(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            val ex = runCatching {
                sf.sql("SELECT nonexistent_column FROM nowhere_table", name = "bad_${UUID.randomUUID().toString().take(6)}")
            }.exceptionOrNull()
            assertTrue(ex is SqlException, "expected SqlException, got $ex")
            assertTrue((ex as SqlException).diagnostics.isNotEmpty())
        }
    }

    @Order(7)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun keyFieldsResolvedFromEngineForLatestBy(transport: Transport) = runBlocking {
        // Wishlist #18: sf.table() with keyFields omitted must read the table's own keyFields
        // (GET /api/tables) instead of a hand-maintained map -- this client never had one, so the
        // proof is that the SAME supersession behavior as supersessionLatestBy above still holds
        // with no keyFields argument at all.
        connectVia(transport).use { sf ->
            val tradeId = "t-${UUID.randomUUID().toString().take(8)}"
            sf.table(EngineFixture.LATEST_TABLE).use { t -> // no keyFields=
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to tradeId, "desk" to "Rates", "notional" to 100.0)))
                t.waitFor(45.seconds) { r -> r.any { it["trade_id"] == tradeId } }

                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to tradeId, "desk" to "Rates", "notional" to 250.0)))
                val rows = t.waitFor(45.seconds) { r ->
                    val match = r.filter { it["trade_id"] == tradeId }
                    match.size == 1 && match.first()["notional"] == 250.0
                }
                assertEquals(1, rows.count { it["trade_id"] == tradeId })
            }
        }
    }

    @Order(8)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun keyFieldsResolvedFromEngineForGroupBy(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            val desk = "Desk-${UUID.randomUUID().toString().take(6)}"
            sf.table(EngineFixture.AGG_TABLE).use { agg -> // no keyFields=
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to "a-${UUID.randomUUID()}", "desk" to desk, "notional" to 40.0)))
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to "b-${UUID.randomUUID()}", "desk" to desk, "notional" to 60.0)))
                val rows = agg.waitFor(45.seconds) { r ->
                    val match = r.filter { it["desk"] == desk }
                    match.size == 1 && match.first()["total"] == 100.0
                }
                assertEquals(1, rows.count { it["desk"] == desk })
            }
        }
    }

    @Order(9)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun keyFieldsResolvedFromEngineForGlobalAggregateStaysOneRow(transport: Transport) = runBlocking {
        // No GROUP BY at all -- engine-resolved keyFields is [] (TableDefinition.KeyFields's "one
        // global group" state, not "no identity"). If the resolver ever collapsed [] to null
        // (whole-row identity) this table would grow a duplicate row on the second push instead
        // of superseding down to exactly one.
        connectVia(transport).use { sf ->
            sf.table(EngineFixture.GLOBAL_AGG_TABLE).use { t -> // no keyFields=
                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to "g-${UUID.randomUUID()}", "desk" to "Global", "notional" to 10.0)))
                t.waitFor(45.seconds) { r -> r.isNotEmpty() }

                sf.push(EngineFixture.SOURCE_NAME, listOf(mapOf("trade_id" to "g-${UUID.randomUUID()}", "desk" to "Global", "notional" to 20.0)))
                val rows = t.waitFor(45.seconds) { r ->
                    r.size == 1 && (((r.first()["trade_count"] as? Number)?.toLong() ?: 0L) >= 2L)
                }
                assertEquals(1, rows.size)
            }
        }
    }

    @Order(10)
    @ParameterizedTest
    @EnumSource(Transport::class, names = ["GRPC", "SIGNALR"])
    @Timeout(90)
    fun adhocSqlRoundtrip(transport: Transport) = runBlocking {
        connectVia(transport).use { sf ->
            val name = "adhoc_roundtrip_${UUID.randomUUID().toString().take(6)}"
            val q = sf.sql(
                "SELECT desk, SUM(notional) AS total FROM ${EngineFixture.LATEST_TABLE} GROUP BY desk",
                name = name,
                keyFields = listOf("desk"),
            )
            try {
                assertTrue(q.ready)
                val listing = sf.adhocTables()
                assertTrue(listing.any { it["name"] == name })
            } finally {
                q.close()
                assertTrue(sf.dropAdhoc(name))
                assertTrue(!sf.dropAdhoc(name)) // already gone
            }
        }
    }
}
